File: app/lib/model/Review.dart
import 'ResourceType.dart';
import 'ReviewStatus.dart';

class Review {
  final String id;
  final String userId;
  final ResourceType resourceType;
  final String resourceId;
  final int rating;
  final ReviewsState state;
  final String? comment;
  final String? moderatedBy;
  final DateTime? moderatedAt;
  final String? moderationReason;
  final DateTime createdAt;
  final DateTime updatedAt;

  Review({
    required this.id,
    required this.userId,
    required this.resourceType,
    required this.resourceId,
    required this.rating,
    this.comment,
    this.moderatedBy,
    this.moderatedAt,
    this.moderationReason,
    required this.createdAt,
    required this.updatedAt,
    required this.state,
  });

  factory Review.fromJson(Map<String, dynamic> json) {
    return Review(
      id: json['id'],
      userId: json['userId'],
      resourceType: ResourceType.values.byName(json['resourceType']),
      resourceId: json['resourceId'],
      rating: json['rating'],
      comment: json['comment'],
      state: ReviewsState.values.byName(json['status']),
      moderatedBy: json['moderatedBy'],
      moderatedAt: json['moderatedAt'] != null
          ? DateTime.parse(json['moderatedAt'])
          : null,
      moderationReason: json['moderationReason'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'resourceType': resourceType.name,
      'resourceId': resourceId,
      'rating': rating,
      'comment': comment,
      'state': state.name,
      'moderatedBy': moderatedBy,
      'moderatedAt': moderatedAt?.toIso8601String(),
      'moderationReason': moderationReason,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  @override
  String toString() {
    return 'Review{id: $id, rating: $rating, resourceType: $resourceType, status: $state}';
  }
}

File: backend/src/main/kotlin/com/veg/bio/infrastructure/table/Review.kt
package com.veg.bio.infrastructure.table

import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.LocalDateTime
import java.util.*

@Entity
@Table(name = "reviews", indexes = [
    Index(name = "idx_review_resource", columnList = "resource_type, resource_id"),
    Index(name = "idx_review_user", columnList = "user_id"),
    Index(name = "idx_review_status", columnList = "status")
])
data class Review(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    val id: UUID? = null,

    @Column(name = "user_id", nullable = false)
    val userId: UUID,

    @Enumerated(EnumType.STRING)
    @Column(name = "resource_type", nullable = false, length = 32)
    val resourceType: ResourceType,

    @Column(name = "resource_id", nullable = false)
    val resourceId: UUID,

    @Column(name = "rating", nullable = false)
    val rating: Int, // 1 à 5

    @Column(name = "comment", length = 1000)
    val comment: String? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 32)
    var status: ReviewStatus = ReviewStatus.PENDING,

    @Column(name = "moderated_by")
    var moderatedBy: UUID? = null,

    @Column(name = "moderated_at")
    var moderatedAt: LocalDateTime? = null,

    @Column(name = "moderation_reason", length = 500)
    var moderationReason: String? = null,

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),

    @UpdateTimestamp
    @Column(name = "updated_at")
    val updatedAt: LocalDateTime = LocalDateTime.now()
) {

    init {
        require(rating in 1..5) { "La note doit être entre 1 et 5" }
        val commentValue = comment
        require(commentValue == null || commentValue.length <= 1000) {
            "Le commentaire ne peut pas dépasser 1000 caractères"
        }
    }

    fun approve(moderatorId: UUID, reason: String? = null) {
        status = ReviewStatus.APPROVED
        moderatedBy = moderatorId
        moderatedAt = LocalDateTime.now()
        moderationReason = reason
    }

    fun reject(moderatorId: UUID, reason: String) {
        status = ReviewStatus.REJECTED
        moderatedBy = moderatorId
        moderatedAt = LocalDateTime.now()
        moderationReason = reason
    }
}

enum class ResourceType {
    RESTAURANT,
    DISH
}

enum class ReviewStatus {
    PENDING,
    APPROVED,
    REJECTED
}